File: day3.hs
import Data.Char (isDigit)


samePrefix :: Eq a => [a] -> [a] -> Bool
samePrefix [] _ = True
samePrefix _ [] = True
samePrefix (x:xs) (y:ys) = x == y && samePrefix xs ys

multiply5 :: String -> Int
multiply5 "" = 1
multiply5 (x:xs) = if x == ')' then 1 else 0

multiply4 :: String -> Int
multiply4 str = read (takeWhile isDigit str) * multiply5 (dropWhile isDigit str)

multiply3 :: String -> Int
multiply3 str = if samePrefix "," str
                then multiply4 $ drop 1 str
                else 0

multiply2 :: String -> Int
multiply2 str = read (takeWhile isDigit str) * multiply3 (dropWhile isDigit str)

multiply :: String -> Int
multiply str = if samePrefix "mul(" str
               then multiply2 $ drop 4 str
               else 0

solveLine :: String -> Int
solveLine "" = 0
solveLine line = multiply line + solveLine (tail line)

solve :: String -> Int
solve = sum . map solveLine . lines 

dropUntilDo :: String -> String
dropUntilDo ('d':'o':'(':')':rest) = rest
dropUntilDo "" = ""
dropUntilDo (x:xs) = dropUntilDo xs

solve2 :: String -> Int
solve2 "" = 0
solve2 line = if samePrefix "don't()" line
                  then solve2 $ dropUntilDo line
                  else multiply line + solve2 (tail line)

main :: IO ()
main = do
    input <- readFile "day3.txt"
    let result = solve input
    print result
    let result2 = solve2 input
    print result2

File: day3.py
import aoc


def is_digit(text: str) -> bool:
    try:
        num = int(text)
        return 0 < num < 1000
    except ValueError:
        return False


def cartes(a: list, b: list) -> list:
    pairs = []
    for first in a:
        for last in b:
            pairs.append((first, last))

    return pairs


def test_substring(sub_str: str) -> bool:
    for num_len_1, num_len_2 in cartes((1, 2, 3), (1, 2, 3)):
        try:
            first_num = sub_str[4:4+num_len_1]
            second_num = sub_str[4+num_len_1+1:4+num_len_1+1+num_len_2]

            correct = sub_str[:4] == 'mul('
            correct &= is_digit(first_num)
            correct &= sub_str[4+num_len_1] == ','
            correct &= is_digit(second_num)
            correct &= sub_str[4+num_len_1+1+num_len_2] == ')'

            if correct:
                return int(first_num), int(second_num)
        except IndexError:
            continue
        
    return (False, False)


def is_do(sub_str):
    try:
        return sub_str[:4] == 'do()'
    except IndexError:
        return False


def is_dont(sub_str):
    try:
        return sub_str[:7] == "don't()"
    except IndexError:
        return False


def solve(lines: list[str]) -> int:
    result = 0
    active = True

    for line in lines:
        for i in range(len(line)):
            sub_str = line[i:]
            if is_do(sub_str):
                active = True
            elif is_dont(sub_str):
                active = False

            if active:
                first, second = test_substring(sub_str)
                if first and second:
                    result += first * second

    return result


if __name__ == '__main__':
    print(solve(aoc.get_input_lines('day3.txt')))